libhb: Add sharpening filters.

bradleysepos opened the pull request:
MPlayer port. CLI only in this patch.

bradleysepos committed:
libhb: Add Unsharp sharpening filter.

Closes #525.

bradleysepos commented:
Working now thanks to John's help. I still need to tune the presets a bit before this goes in; right now they are a blind guess.

bradleysepos commented:
Unsharp presets and tunes have been tweaked.

I believe this is ready to be merged, pending everyone's approval. GUIs can be updated to add support whenever people have time. Should be nearly identical to implementing NLMeans in the GUIs.

jstebbins commented on libhb/preset.c:
```diff
@@ -1371,6 +1371,60 @@ int hb_preset_apply_filters(const hb_dict_t *preset, hb_dict_t *job_dict)
+    sharpen = hb_value_type(sharpen_value) == HB_VALUE_TYPE_STRING ? (
+        !strcasecmp(hb_value_get_string(sharpen_value), "off") ? 0 :
+        !strcasecmp(hb_value_get_string(sharpen_value), "unsharp") ? 1 : 2) :
+        hb_value_get_int(sharpen_value);
```
Is there a reason we would want to allow int values for this?  I supported int values for some things in order to support legacy values.  But going forward, I think we should discourage it.

jstebbins commented on libhb/unsharp.c:
```diff
@@ -0,0 +1,261 @@
+    {
+        if (y < height) src2 = src;
+        memset(SR, 0, sizeof(SR[0]) * (2 * steps -1 ));
+        for (x=-steps; x<width+steps; x++)
```
Add white space please

jstebbins commented on libhb/unsharp.c:
```diff
@@ -0,0 +1,261 @@
+                Tmp2 = SC[z+0][x+steps] + Tmp1; SC[z+0][x+steps] = Tmp1;
+                Tmp1 = SC[z+1][x+steps] + Tmp2; SC[z+1][x+steps] = Tmp2;
+            }
+            if (x>=steps && y>=steps)
```
white space

jstebbins reviewed (commented):
How did you envision the UI for this looking.  For filters that only have one type (e.g. detelecine), we've only been providing a widget for what is essentially the preset for that filter.  And for filters that have multiple types (e.g. denoise), we provide a widget for filter type and preset (plus tune in some cases).

You've povided keys that correspond to type, preset, tune, and custom.  So the keys align with the multiple type paradigm, but the UI would look a little strange with just "Off" and "Unsharp" type options.  I'm thinking that for now, we should provide an "Unsharp" filter in the UI with only preset, tune, and custom widgets, where the "preset" values include an "Off" option (i.e. no type selection).  The UI code would be responsible for knowing that "preset" value "Off" means type "Off".

bradleysepos commented:
I envisioned the UI looking similar to Denoise. It allows us to identify the filter type and filter ID separately; Unsharp masking is a type of Sharpening. I can see adding other sharpening filters down the line, if it proves useful.

bradleysepos commented:
To be clear, I think it would be a mistake form a UX perspective to simply have "Unsharp" without an explanation. Most lay people do not know that unsharp masking is a type of sharpening. Detelecine is by nature more technical.

bradleysepos commented:
Pushed two commits to address comments. Thanks for the feedback.

jstebbins commented:
So call it the "Sharpen" instead of "Unsharp" in the UI.  I still think it would be kind of weird to have a type "Sharpen Filter" dropdown whos only options are "Off" and "Unsharp".  I've already done an implementation for LinGui that labels the filter "Sharpen Filter" and has preset and tune dropdowns and a custom entry box.  In the preset box there is an "Off" option.  If in the future we add another sharpen filter type, it's trivial to add the dropdown for it.

bradleysepos commented:
If in the future... meaning in a few days? ;)

bradleysepos committed:
libhb: Add LapSharp sharpening filter.

bradleysepos commented:
Rebased and squashed. Added second sharpening filter, LapSharp. It's functional to test.

I still need to do some fancy work with LapSharp to make it able to switch between multiple convolution kernels of different sizes. @jstebbins How would you go about doing this, use a dynamic array?

Also need to switch LapSharp's kernel selection from int to string value.

jstebbins commented:
I'm not entirely sure what you mean by multiple kernels, but I'll take a stab.  If my guess is right, I would start by not using 2 dimensional arrays for the kernel. Make them single dimensional and do the indexing in your arithmetic.  e.g.
> pixel += kernel[((j - offset_min) * kernel_size) + k - offset_min] * *(src + stride * (y + j) + (x + k));

Doing the above will allow you to statically declare your kernels something like this:
```
typedef struct {
    const int * kernel;
    const int kernel_size;
} kernel_t;

    static const int kernel1[] = { 0,  0, -1,  0,  0,
                                   0, -1, -2, -1,  0,
                                  -1, -2, 24, -2, -1,
                                   0, -1, -2, -1,  0,
                                   0,  0, -1,  0,  0};
    static const int kernel2[] = { 0,  0, -1,  0,  0, 0,
                                   0, -1, -2, -1,  0, 0,
                                  -1, -2, 24, -2, -1, 0,
                                   0, -1, -2, -1,  0, 0,
                                   0, -1, -2, -1,  0, 0,
                                   0,  0, -1,  0,  0, 0};
    static kernel_t kernels[] = {
        { kernel1, 5 },
        { kernel2, 6 }};
```

bradleysepos commented:
Thanks @jstebbins, your advice was spot on. If you wouldn't mind making sure I haven't missed a free() related to these changes, it should be ready to go.

jstebbins commented on libhb/param.c:
```diff
@@ -408,6 +455,294 @@ static hb_dict_t * generate_nlmeans_settings(const char *preset,
+        !strcasecmp(preset, "stronger"))
+    {
+        double strength[2];
+        char  *kernel_string[2];
```
Make this const and drop the strdups.  This currently leaks.

jstebbins commented on libhb/lapsharp.c:
```diff
@@ -0,0 +1,280 @@
+typedef struct
+{
+    double strength;       // strength
+    char  *kernel_string;  // which kernel to use (user string)
```
This is only used in init.  I would make it a local variable char * kernel_string[3] where it is used

jstebbins commented on libhb/lapsharp.c:
```diff
@@ -0,0 +1,280 @@
+    {
+        hb_dict_t * dict = filter->settings;
+        hb_dict_extract_double(&pv->plane_ctx[0].strength,      dict, "y-strength");
+        hb_dict_extract_string(&pv->plane_ctx[0].kernel_string, dict, "y-kernel");
```
hb_dict_extract_string allocates memory.  kernel_string will need to be freed.

bradleysepos commented:
Comments addressed, thanks.

jstebbins commented on libhb/param.c:
```diff
@@ -408,6 +455,292 @@ static hb_dict_t * generate_nlmeans_settings(const char *preset,
+        else
+        {
+            fprintf(stderr, "Unrecognized lapsharp tune (%s).\n", tune);
+            free(kernel_string);
```
Invalid free

jstebbins commented on libhb/param.c:
```diff
@@ -408,6 +455,292 @@ static hb_dict_t * generate_nlmeans_settings(const char *preset,
+        hb_dict_set(settings, "cb-strength", hb_value_double(strength[1]));
+        hb_dict_set(settings, "cb-kernel",   hb_value_string(kernel_string[1]));
+
+        free(kernel_string);
```
Invalid free

jstebbins commented:
After fixing those 2 "free" issues locally, tests look good and code looks good.

bradleysepos commented:
Merged, thanks again for the feedback. 😸
